ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -0,0 +1,558 @@
+-- dense chain would have to lose at least one race within the Genesis window.
+checkAdversarialChain ::
+  forall base hon adv.
+     AdversarialRecipe base hon
+  -> ChainSchema base adv
+  -> Exn.Except (AdversarialViolation hon adv) ()
+checkAdversarialChain recipe adv = do
+    checkStart
+    checkCount
+    checkRaces
+  where
+    AdversarialRecipe {
+        arHonest = ChainSchema winH vH
+      ,
+        arParams = (Kcp k, Scg s, Delta d)
+      ,
+        arPrefix
+      } = recipe
+
+    ChainSchema winA vA = adv
+
+    checkStart = do
+        let startA       = C.windowStart winA :: C.Index base SlotE
+            intersection = startA C.- 1       :: C.Index base SlotE
+
+        -- The intersection must be at an active slot in the honest chain.
+        case C.toWindow winH intersection of
+            Nothing -> do
+              -- the genesis block is the only permissible anchor outside of @hon@
+              when (startA   /= C.Count 0) $ Exn.throwError $ BadAnchor HonestActiveMustAnchorAdversarial
+              when (arPrefix /= C.Count 0) $ Exn.throwError $ BadAnchor WrongNumberOfHonestPredecessors
+
+            Just i -> do
+                let _ = i :: C.Index hon SlotE
+
+                when (BV.testV S.inverted vH i) $ do
+                    Exn.throwError $ BadAnchor HonestActiveMustAnchorAdversarial
+
+                C.SomeWindow Proxy precedingSlots <-
+                    pure $ C.withWindowBetween (C.windowSize winH) (C.Lbl @"foo") (C.Count 0) i
+                let pc = BV.countActivesInV S.notInverted (C.sliceV precedingSlots vH)
+
+                -- arPrefix must correctly count the active slots in the part of
+                -- the chain upto the intersection
+                when (C.fromWindowVar precedingSlots (C.toVar pc) /= arPrefix) $ do
+                    Exn.throwError $ BadAnchor WrongNumberOfHonestPredecessors
+
+    checkCount = do
+        let pc = BV.countActivesInV S.notInverted vA
+        when (C.toVar pc <= 0) $ Exn.throwError BadCount
+
+    -- the youngest slot in which the adversarial schedule cannot have accelerated
+    --
+    -- (IE s past the first active adversarial slot, or delta passed the k+1 slot
+    --  after the intersection)
+    youngestStableA :: C.Index base SlotE
+    youngestStableA =
+        let sYoungest = case BV.findIthEmptyInV S.inverted vA (C.Count 0) of
+              BV.JustFound firstActiveA ->
+                -- if s=0, then the slot of their block is the youngest stable slot
+                C.fromWindow winA $ firstActiveA C.+ s
+              BV.NothingFound           ->
+                -- the rest of the function won' force this since there are no
+                -- adverarial active slots
+                error "dead code"
+            kPlus1stYoungest = case BV.findIthEmptyInV S.inverted vH (C.toIndex arPrefix C.+ k) of
+              BV.JustFound kPlus1st -> C.fromWindow winH $ kPlus1st C.+ d
+              BV.NothingFound       -> C.Count d
```

Comment:
This case seems wrong to me. It would only be accurate if `k+1` were `0`, wouldn't?

I think `C.windowLast winH` would be better. Still inaccurate, but I think it'd at least lead to the right behavior.

Accurate would be: `C.windowLast winH C.d (k+1 - x) C.+ d` where `x` is how many blocks the honest chain does have remaining (where `x < k +1` since we're not in the `JustFound` alterantive).

Reply:
Agreed and adopted in e1135c0.

Reply:
I pushed d2de74c37b8e45b4a64064034af4f2470cbd6f0c, which I think does the "exact" calculation and isn't too onerous.